### ISSUE TITLE
Add an option lock_wait to the apt module

### ISSUE THREAD

**Issue by drym3r** [ansible/ansible#25414]
##### ISSUE TYPE
 - Feature Idea 
This is a copy of the issue on the old repository: https://github.com/ansible/ansible-modules-core/issues/4717
There was some work on it, I don't know why it's not added: https://github.com/ansible/ansible-modules-core/pull/4842

##### COMPONENT NAME
apt

##### SUMMARY
This is a feature request but also works around a problem that is very common when manipulating apt packages/repositories on a given host from multiple instances of apt.

Currently if there is an instance working with apt (via ansible apt module or not) there is a lock which is being held exclusively and as such no other apt instance can get a hold of. The result is that if there's a second instance of apt which tries to acquire the lock it gets an instant failure. In the case of doing this via ansible the result is:

msg: Failed to lock apt for exclusive operation

This is the default apt behavior and as far as I can tell there is not way to pass a flag to apt command line, or apt configuraion, to tell it to wait for the lock to be released and then only timeout after a certain period of time.

It would be very useful if ansible's apt module itself was to provide such a feature. Say an optional flag - for example lock_wait=60 which would make ansible try to acquire the lock every second for X seconds and eventually timing out. This way if two ansible playbooks running at the same time are trying to use the apt repository they will have much higher chance of succeeding.

Currently one of them errors out only so that the user retry the entire playbook .

##### STEPS TO REPRODUCE
Run two ansible instances each utilizing the apt module at the very same time - one of them will fail with the above error.

##### EXPECTED RESULTS
If we have this feature in ansible then we don't need to wrap each individual playbook's apt calls with retries and have it built-in the apt module itself.

**Comment by lyz-code** [comment]
It also happens to me, but instead of two ansibles, when you try to execute a playbook and the unattended upgrades is on (by default for Ubuntu 16.04).

**Comment by ansibot** [comment]
cc @mgwilliams
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by brainstorm** [comment]
It would help if not only `apt` benefited from this feature, I guess that other package managers also have some sort of locking mechanism... perhaps abstracting it away to [package module](https://docs.ansible.com/ansible/latest/package_module.html) instead would be more portable/beneficial in general?

**Comment by ITBlogger** [comment]
Any updates on this? Still having the issue with 2.5.1

**Comment by mbuiro** [comment]
Literally just encountered this problem while working on deploying lxd containers, this is really causing some issues for me!

**Comment by cledoux** [comment]
There's a workaround listed here: https://github.com/ansible/ansible/issues/16593#issuecomment-253026793

Basically, use the shell module to wait for the lock file to be empty.

```
- name: Wait for automatic system updates
  become: yes
  shell: "while fuser /var/lib/dpkg/lock >/dev/null 2>&1; do sleep 1; done;"
```

**Comment by brainstorm** [comment]
@cledoux In my experience, that does not work reliably since `fuser` only checks for a file being used by a process, not "being empty"... I've targeted unattended upgrades myself since that's what was the problem for me, but I hope it can be extrapolated to dpkg as well:

```
# Does not work since the lock file never gets removed, actually
#- name: Wait for automatic system updates to finish first
#  wait_for:
#    path: /var/lib/dpkg/lock
#    state: absent
#    timeout: 600

# https://github.com/ansible/ansible/issues/25414 ... unreliable :/
#- name: Wait for automatic system updates
#  become: true
#  become_user: root
#  shell: while fuser /var/lib/dpkg/lock >/dev/null 2>&1; do sleep 1; done;

# Trying target unattended-upgrades, works every time
- name: Wait for automatic system updates
  shell: while pgrep unattended; do sleep 10; done;
```

**Comment by lisp-ceo** [comment]
This should integrate with systemd

```
- name: Wait for any possibly running unattended upgrade to finish
  raw: systemd-run --property="After=apt-daily.service apt-daily-upgrade.service" --wait /bin/true
```

**Comment by dagwieers** [comment]
Just as a reference, we are looking into file-locking to ensure concurrent (write) access works better in Ansible, see #47322. Once we have a solution for this, we could leverage this mechanism (and hopefully use apt's own lock file, or if not, some other file).

**Comment by nyetwurk** [comment]
I had to do something like this (yes, using `fuser`, not looking for the `unattended` process)

Since sometimes `lock-frontend` is held but `lock` isn't:
```
- name: Wait for automatic system updates
  become: true
  shell: while sudo fuser /var/lib/dpkg/{{ item }} >/dev/null 2>&1; do sleep 1; done;
  with_items:
    - lock
    - lock-frontend
```
Note that this workaround is still race prone, since if another apt starts between this and the upcoming ansible apt, the ansible apt one will still fail.

**Comment by minusdavid** [comment]
> # Trying target unattended-upgrades, works every time
> - name: Wait for automatic system updates
>   shell: while pgrep unattended; do sleep 10; done;
> ```

This creates an infinite loop if the unattended-upgrades service is running in the background (like in Ubuntu 18.04).

It looks like apt-daily and apt-daily-upgrade both separately run unattended-upgrades at boot, so there is also a race condition between the two services where your pgrep will return a positive integer.

**Comment by minusdavid** [comment]
> This should integrate with systemd
> 
> ```
> - name: Wait for any possibly running unattended upgrade to finish
>   raw: systemd-run --property="After=apt-daily.service apt-daily-upgrade.service" --wait /bin/true
> ```

This is perfect. Waiting for apt-daily.service and apt-daily-upgrade.service was exactly what I wanted to do. 

Checking for unattended-upgrades itself is problematic since both services call it separately. Checking for the lock file is problematic for a similar reason. Trying to do provision time changes to unattended upgrade configuration is mostly too late. Just checking for the dpkg lockfile doesn't take into account the apt list lockfile.

This is a great solution when provisioning I reckon.

**Comment by p-thurner** [comment]
I just found the tool `aptdcon` - from `man aptdcon`: "aptdcon allows to perform package management tasks, e.g. installing or removing software, using aptdaemon."

I just checked on my Ubuntu Laptop:
- I can run `aptdcon --install atop` and `aptdcon install tmux` in two shells at the same time and it queues nicely
- I can run another pkg manager like `apt-get remove atop` and then in a second shell `aptdcon -i tmux` and aptdcon will reply "waiting for another pkg manager: apt-get`
- I just checked and it also waits for a `dpkg --install foobar.deb`

Seems like this feature already exists. Simplest way of using it in ansible would be sth like:
```yaml
- name: install something
  shell: aptdcon --install atop tmux
```

This also enables using apt in multiple tasks parallel with ansible.

I assume writing a Ansible module for this aptdcon utility would be a nice idea.

**Comment by p-thurner** [comment]
So after searching a bit I found that packagekit: https://www.freedesktop.org/software/PackageKit/

brings a tool called `pkcon` that also features a queue:
```
root@host:~# pkcon -y install kate
Resolving                     [=========================]         
Installing                    [=========================]         
Waiting in queue              [=========================]         <-- Here it waited for another pkcon
Loading cache                 [=========================]         
Downloading packages          [=========================]         
Running                       [=========================]         
Installing packages           [=========================]         
Finished                      [=========================] 
```
I just tried and it also waits for:
- `apt install tmux`
- `dpkg --install foo.deb`

Here is its output in that case:
```
root@x270:~# pkcon -y remove kate
Resolving                     [=========================]         
Removing                      [=========================]         
Waiting for package manager lock [=========================]   <-- Waiting for a `apt install` command in another shell here
```

This looks like a good solution for Debian 9 and Ubuntu 18.04.

**Comment by nyetwurk** [comment]
> 
> 
> This should integrate with systemd
> 
> ```
> - name: Wait for any possibly running unattended upgrade to finish
>   raw: systemd-run --property="After=apt-daily.service apt-daily-upgrade.service" --wait /bin/true
> ```

This also worked perfectly for me. It is still prone to race of course but it solved my immediate issues.

**Comment by nafallo** [comment]
aptdcon is just a python script that imports aptdaemon. We should likely rewrite the apt_module to try aptdaemon first and fall back to python{,3}-apt instead...

**Comment by nafallo** [comment]
aptdaemon is on it's death bed, apparently. I suppose we could start with a small while loop under the apt-cache-LockFailedException and retry running the main function a few times?

**Comment by gentoo9ball** [comment]
Any more luck figuring out a more permanent workaround?

**Comment by ScottG489** [comment]
The workaround in https://github.com/ansible/ansible/issues/25414#issuecomment-440549135 is working for me but I wanted to call out that this issue is still present.

**Comment by bcoca** [comment]
fixed via #69284

instead of an apt specific fix, its a generic task timeout.

**Comment by madeddie** [comment]
> fixed via #69284
> 
> instead of an apt specific fix, its a generic task timeout.

How does this fix the issue? The problem is the task fails because it cannot get a lock, this was never a timeout issue. It fails immediately because aptitude or apt fail to get a lock, they don't retry or wait for a lock.

**Comment by nyetwurk** [comment]
Exactly this.

**Comment by stas-sbi** [comment]
Here is what we use as workarounds:
```
- name: APT LISTS - wait for the lock to be released
  raw: 'sudo lslocks -uo path | grep -E "^(PATH|/var/lib/apt/lists/lock)$"'
  register: result
  until: result.stdout|trim == "PATH"
  retries: 100
  delay: 5
```
```
- name: DPKG - wait for the lock to be released
  raw: 'sudo lslocks -uo path | grep -E "^(PATH|/var/lib/dpkg/lock|/var/lib/dpkg/lock-frontend)$"'
  register: result
  until: result.stdout|trim == "PATH"
  retries: 300
  delay: 5
```

**Comment by giner** [comment]
Here is what we use as workarounds:
```
- name: APT LISTS - wait for the lock to be released
  raw: 'sudo lslocks -uo path | grep -E "^(PATH|/var/lib/apt/lists/lock)$"'
  register: result
  until: result.stdout|trim == "PATH"
  retries: 100
  delay: 5
```
```
- name: DPKG - wait for the lock to be released
  raw: 'sudo lslocks -uo path | grep -E "^(PATH|/var/lib/dpkg/lock|/var/lib/dpkg/lock-frontend)$"'
  register: result
  until: result.stdout|trim == "PATH"
  retries: 300
  delay: 5
```

**Comment by nyetwurk** [comment]
Unfortunately all of the workarounds above still have race issues, because test/get isn't atomic.

**Comment by madeddie** [comment]
@giner yeah a similar solution has been posted above, but the original request was for some mechanism to have apt module wait for the lock itself. This issue should not be closed.

**Comment by nyetwurk** [comment]
@madeddie Not just wait, but atomic lock acquisition to avoid lock races.

**Comment by ScottG489** [comment]
@bcoca Based on the above response, could we please have this reopened?

**Comment by shitalp** [comment]
could we please have this reopened? I am completely stucked due to this. seeking for help.

**Comment by nate-duke** [comment]
There's a chance we're shouting at the void here if Redhat is not monitoring closed issues. Perhaps we that are still interested in this issue being resolved should file a new issue requesting that this one be reopened.

Worth mentioning that this issue dates back to 2016 so if you're seeking immediate relief it's best to explore the various workarounds in addition to monitoring this one.

FWIW the workround we use is to "try again later" as that's the only way we've found to keep our playbooks progressing.

**Comment by bcoca** [comment]
k, misread the issue, its not a 'waiting for timeout' but 'retry on failed lock'

**Comment by bcoca** [comment]
In any case, isn't this what until/retry handles in a generic way?

You should not need to add this feature to each action.

needs_info

**Comment by nyetwurk** [comment]
A concurrent dpkg/apt action can take a very long time. Not being able to distinguish a real error from a lock conflict in a sane amount of time is extremely problematic.

**Comment by bcoca** [comment]
the lock currently fails immediately, what the ticket is asking is for 'retry on lock' which the task i posted above does in a configurable manner.

**Comment by shitalp** [comment]
Yeah, Same issue i m also facing i need solution for that.

**Comment by shitalp** [comment]
Even i tried with this logic wont work for me as sometimes it works sometimes it doesnot.

- name: Wait for automatic system updates
  become: true
  shell: while sudo fuser /var/lib/dpkg/{{ item }} >/dev/null 2>&1; do sleep 1; done;
  with_items:
    - lock
    - lock-frontend

**Comment by ScottG489** [comment]
I've found this bug report for a wait feature in apt: **[apt: support option to wait for lockfile](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=754103)**

It was closed (Feb 26, 2020) and the wait feature is available in apt 1.9.11 forward. I've tested it out on my local machine (apt 2.0.2ubuntu0.1) and it seems to work at least **for installs**. This is the output you would receive (with this line repeating every second until the lock was released):

```
Waiting for cache lock: Could not get lock /var/lib/dpkg/lock-frontend. It is held by process 3899676 (apt)
```

I'm not sure if the wait time is configurable though or if there is even a default. I let the wait run for 1000 seconds before killing it.

A few important things to point out. First this only works for `apt`, not `apt-get`. Also importantly this doesn't seem to work with `apt update`. It seems that `apt update` can run in parallel to an install. However, having two parallel instances of `apt update` will still fail immediately with no wait.

So personally this wouldn't fix my issue because I had playbooks failing because of contention on whatever lock `apt update` uses. However, I didn't see this mentioned in here anywhere and thought I would call it out in case it could help someone else out. Though I imagine this feature would still have to be integrated into ansible regardless.

**Comment by ansibot** [comment]
@drym3r This issue is waiting for your response. Please respond or the issue will be closed.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by nyetwurk** [comment]
Do not close this bug please.

**Comment by nyetwurk** [comment]
What info do we need here? Just failing doesn't work, we need a retry. A simple retry doesn't work because it isn't atomic and is full of race issues.

**Comment by drym3r** [comment]
Not sure what @ansibot is expecting from me, but I agree the issue is not solved.

**Comment by ansibot** [comment]
<details><summary><b>Components</b></summary>

[lib/ansible/modules/apt.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/apt.py)
support: core
maintainers: mgwilliams

</details>

<details><summary><b>Metadata</b></summary>

**waiting_on**: maintainer
[needs_info](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md#label-needs_info): False
</details>

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by JonTheNiceGuy** [comment]
Just worth noting, the stanza I'm using is:

```yaml
- name: Some apt action
  apt:
    name: "{{ item }}"
    state: present
  loop: "{{ install_these_packages }}"
  register: apt_action
  retries: 100
  until: apt_action is success or ('Failed to lock apt for exclusive operation' not in apt_action.msg and '/var/lib/dpkg/lock' not in apt_action.msg)
```

My intention is to update the apt module's documentation with this stanza.

**Comment by samdoran** [comment]
When this feature is implemented, the documentation change in https://github.com/ansible/ansible/pull/73079 needs to be removed.

**Comment by thom-vend** [comment]
> This should integrate with systemd
> 
> ```
> - name: Wait for any possibly running unattended upgrade to finish
>   raw: systemd-run --property="After=apt-daily.service apt-daily-upgrade.service" --wait /bin/true
> ```

Hi there 😊 ,
I've been using this for a while, I got my ansible task stuck on it because the `apt-daily.service` was `dead` and `inactive` from the `systemctl status apt-daily.service`. Note `apt-daily.service` is a `Type=oneshot` `RemainAfterExit=no` on Ubuntu 20.04 LTS.

I'm looking for a work around, first idea is to invoke it via the timeout command to allow 5 minutes max but still not good if the APT unit need more time to complete. 🤔 🤔 🤔

**Comment by lilatomic** [comment]
I think this issue will be fixed by #74095

**Comment by samdoran** [comment]
Resolved by #74095.
